### ISSUE TITLE
Feature suggestion: Loss Normalization

### ISSUE THREAD

**Issue by Hvass-Labs** [tensorflow/tensorflow#6414]
I would like to suggest the following feature be added to TensorFlow. As far as I know, the feature does not exist in TensorFlow. There are some issues that must be discussed and solved before making the pull-request.

### Motivation

When optimizing multiple objectives in TensorFlow, the combined loss-function is typically a weighted sum of the individual loss-functions. Finding the weights for the loss-functions is done experimentally as the weights may depend on a number of factors.

For example, in Style Transfer we want to combine two images so the result has the 'style' of one input image and the 'content' of another image. We do this by creating two loss-functions that we will optimize together. The problem is that when we choose different layers in the neural network for the loss-functions, then their magnitudes may change dramatically, so the weights for the loss-functions will also need to be modified.

The solution I came up with in my recent tutorial on Style Transfer, is to automatically normalize each loss-function so we can define the weights independently of any given choice of layers:

https://github.com/Hvass-Labs/TensorFlow-Tutorials/blob/master/15_Style_Transfer.ipynb

The solution I propose below is a bit more elegant than in the tutorial, and I think it can be made even better.

### Overall Idea

The idea is to evaluate each loss-function to its scalar-value and then divide the loss-function with this scalar-value. This creates a new node in the TensorFlow graph which holds the original loss-function divided by some scalar-value. The new node in the graph is still a mathematical function that can be differentiated, and for the given input it evaluates to the value 1. We do this normalization for each of the loss-functions so they all evaluate to 1. Then we can define loss-weights more easily and independently of e.g. the exact choice of layers in the Style Transfer algorithm.

### Code

In terms of pseudo-code, what we want to do is simply:

    def loss_normalize(loss):
        return loss / value_of(loss)

But if we merely return `loss/loss` then it does not calculate the value of `loss` for the denominator, instead we get a mathematical function divided by itself. I also tried `tf.identity(loss)` but that did not work either. Neither does type-casting.

So we need a small hack. We create a scalar float-variable and assign the loss-value to it. When the variable is used in an assignment like this, then it is not necessary to initialize it first. This works:

    def loss_normalize(loss):
        loss_value = tf.Variable(1.0)
        
        loss_normalized = loss / loss_value.assign(loss)

        return loss_normalized

### Usage

Imagine that we have defined two loss functions, e.g.:

    style_loss = ...
    content_loss = ...

Normally we would have some strange weights that might need to be changed if we choose different layers in the network for the loss-functions. These weights do not indicate how we are actually weighting the loss-functions for the style and content, for example:

    combined_loss = 1e-10 * style_loss + 1e-3 * content_loss

If instead we normalize the loss-functions first, then we can define the weights more intuitively. For example, we might want 90% of the combined-loss to be for the style and only 10% for the content. We would do it like this:

    combined_loss = 0.9 * loss_normalize(style_loss) + \
                    0.1 * loss_normalize(content_loss)

Even if you change the layers used in the two loss-functions, we do not need to change our weights. This may be useful in many other applications than just Style Transfer.

### Issues

Before I make a pull-request there are several issues that must be discussed and solved.

1) The hack using a variable to get the loss-value is not very elegant. I wonder if it would be possible to add something like `value_of(loss)` to TensorFlow? I guess it would be a bit similar to `loss.eval()` except that we don't want to run the session.

2) Division by zero should be handled somehow. If the loss-function is inherently non-negative then we can add a small number such as 1e-10 to the denominator. But this would not be bomb-proof for loss-functions that can take on negative values. Is there a good standard solution for this in TensorFlow?

3) It might be a better design to make `loss_normalize()` take a list of loss-functions so we don't have to wrap all of the loss-functions individually. What do you think? Is there an elegant way of implementing this in TensorFlow, or do I have to resort to Python's list comprehension?

4) It might be useful to have a placeholder bool that decides whether or not to update the normalization. This would allow us to normalize the loss-functions only in the first iteration, or we could normalize every n'th iteration. The reason this might be useful, is that the optimizer might get trapped in a local optimum if we normalize the loss-functions in every iteration. I experimented a bit with `tf.placeholder_with_default()` and `tf.cond()` but it gave me some strange errors about tensor-shapes. Any ideas on how to make an elegant implementation for this?

**Comment by aselle** [comment]
Seems like learn might be a good place for this? Could you two @ilblackdragon  and @martinwicke  comment on your opinion of the feature. Thanks!

**Comment by ilblackdragon** [comment]
@Hvass-Labs You current implementation of `loss_normalize` will always return 1.0 if losses are already scalars [which they usually are after calling for example `tf.losses.*`]:
```
In [8]: def loss_normalize(loss):
        loss_value = tf.Variable(1.0)
        return loss / loss_value.assign(loss)

In [10]: with tf.Session() as sess:
    print(sess.run(loss_normalize(1.5)))

1.0
```

If you are just trying to normalize for example a vector or matrix of raw errors [for example `tf.abs(a - b)` part of `tf.reduce_mean(tf.abs(a - b))` - the absolute mean error loss], you can use one of https://www.tensorflow.org/api_docs/python/nn/normalization tools.  

If I understood your Ipython Notebook - you are suggesting to use normalization by dividing on previous value of this tensor. You can add it to the tools in the Normalization list - this doesn't need to be loss specific. To perform updates at correct time add the assign op to `GraphKeys.UPDATE_OPS`, which will be auto-added to list of ran tensors by `optimize_loss`. 

Also consider using tf.nn.moments and tf.nn.normalize_moments as it may be a better estimation of what you are trying to do.

Let me know if this makes sense to you!

**Comment by Hvass-Labs** [comment]
Thanks for the detailed comments.

The Notebook for my Tutorial #15 has a slightly awkward implementation of this idea. The implementation I wrote above is cleaner.

However, from your comments I can see that the idea is still not perfectly clear. I think it would help if I write a Python Notebook to demonstrate it, and then perhaps you can give suggestions on where it might belong in TensorFlow and possibly how to improve the implementation.

But I'm currently busy doing some research related to TensorFlow, so I'll have to come back to this thread later. It will probably be several weeks or more.

**Comment by Hvass-Labs** [comment]
I'm sorry for the delay, I was doing some research that dragged out.

When I proposed this feature I thought it might be useful in many other scenarios. But after more consideration, I don't think the idea is sufficiently general and mature at this stage to be included in TensorFlow core.

I have made a Python Notebook which demonstrates the idea I had in mind:

https://gist.github.com/Hvass-Labs/3744f05fa05d6ef82dc873c4c32737ef

Try switching the bool `use_loss_normalization` and see what happens to the resulting image at the bottom. However, this is not the best example because the two loss-functions are fairly close in value.

This type of loss-normalization would work better for loss-functions that are several orders of magnitude in difference, where one loss-function would tend to dominate, and whenever you change something in the configuration of the problem, the relative values of the loss-functions would change so you also needed to change the weights. In such cases, automatic loss-normalization would be useful.

Perhaps someone in the future might find the idea useful and could build on this. This type of loss-normalization might make a nice little research paper if someone is interested in doing it (I don't have time and I have many other ideas to work on).
